Source/LinqToDB/DataProvider/Firebird/FirebirdSqlOptimizer.cs

```diff
@@ -343,6 +297,47 @@ private SqlStatement WrapParameters(SqlStatement statement, EvaluationContext co
 			return statement;
 		}
 
+		private static bool CastRequired(IReadOnlyList<IQueryElement> parents)
+		{
+			for (var i = parents.Count - 1; i >= 0; i--)
+			{
+				// went outside of subquery, mission abort
+				if (parents[i] is SelectQuery)
+					return false;
+
+				// part of select column
+				if (parents[i] is SqlColumn)
+					return true;
+
+				// part of output clause
+				if (parents[i] is SqlOutputClause)
+					return true;
+
+				// insert or update keys used in merge source select query
+				if (parents[i] is SqlSetExpression set
+					&& i == 2
+					&& (parents[i - 1] is SqlInsertClause || parents[i - 1] is SqlUpdateClause)
+					&& parents[i - 2] is SqlInsertOrUpdateStatement insertOrUpdate
```

Comment:
Since `i==2` per line 318, simplify constants here:

```suggestion
					&& (parents[1] is SqlInsertClause || parents[1] is SqlUpdateClause)
					&& parents[0] is SqlInsertOrUpdateStatement insertOrUpdate
```